plugins/osgi/src/main/java/org/apache/struts2/osgi/BundlePackageLoader.java

```diff
@@ -45,24 +47,96 @@
 
     private static final Logger LOG = LogManager.getLogger(BundlePackageLoader.class);
 
+    private Container contextContainer = null;
+
+    @Override
     public List<PackageConfig> loadPackages(Bundle bundle, BundleContext bundleContext, ObjectFactory objectFactory,
                                             FileManagerFactory fileManagerFactory, Map<String, PackageConfig> pkgConfigs) throws ConfigurationException {
+        if (pkgConfigs == null) {
+            throw new IllegalArgumentException("Cannot load packages from a null package configuration");  // Better than a NPE.
+        }
+
         Configuration config = new DefaultConfiguration("struts.xml");
+
+        LOG.trace("LoadPackages - After config constructed.  Before BundleConfigurationProvider constructed");
+
         BundleConfigurationProvider prov = new BundleConfigurationProvider("struts.xml", bundle, bundleContext);
+
+        LOG.trace("LoadPackages - After BundleConfigurationProvider constructed.  Before config.addPackageConfig loop");
+
         for (PackageConfig pkg : pkgConfigs.values()) {
             config.addPackageConfig(pkg.getName(), pkg);
         }
+
+        LOG.trace("LoadPackages - After config.addPackageConfig loop.  Before prov.setObjectFactory()");
+
         prov.setObjectFactory(objectFactory);
+
+        final DefaultFileManagerFactory defaultFileManagerFactory = new DefaultFileManagerFactory();
+        final Container container = getContextContainer();
+
+        if (container == null) {
+            LOG.warn("LoadPackages - Config Container is null.  May cause a NPE to be thrown");
+        }
+
+        defaultFileManagerFactory.setContainer(container);
+
+        if (fileManagerFactory == null || fileManagerFactory.getFileManager() == null) {
+             LOG.warn("LoadPackages - FileManagerFactory parameter is null or produces a null FileManager, replacing with a new DefaultFileManagerFactory instance");
+             fileManagerFactory = defaultFileManagerFactory;
+        }
+
         prov.setFileManagerFactory(fileManagerFactory);
```

Comment:
I would change this a bit because right now you are re-assigning to the method's parameter, also you are creating `DefaultFileManagerFactory` upfront even if it could be not needed. What about such change:

```java
if (fileManagerFactory == null || fileManagerFactory.getFileManager() == null) {
  LOG.warn("LoadPackages - FileManagerFactory parameter is null or produces a null FileManager, replacing with a new DefaultFileManagerFactory instance");

  final DefaultFileManagerFactory defaultFileManagerFactory = new DefaultFileManagerFactory();
  final Container container = getContextContainer();

  if (container == null) {
    LOG.warn("LoadPackages - Config Container is null.  May cause a NPE to be thrown");
  } else {
    container.inject(defaultFileManagerFactory);
  }

  prov.setFileManagerFactory(defaultFileManagerFactory);
} else {
  prov.setFileManagerFactory(fileManagerFactory);
}
```

Reply:
I cannot remember why I re-used the `fileManagerFactory` parameter (probably trying to limit the number of lines that I was changing at the time).  Anyway, your recommended changes above make more sense, so we will go with that logic instead (in an updated commit).